### ISSUE TITLE
[TODO] Ensure MIDI 1.0 APIs (WinMM / WinRT) have WinMM-compatible or better naming for devices/ports under new enumeration code.

### ISSUE THREAD

**Issue by Psychlist1972** [microsoft/MIDI#114]
Many existing editors and configuration apps rely on an exact port name as they appear in WinMM MIDI. We will need to ensure our new names for MIDI 1.0 devices through the MIDI 1.0 APIs are consistent with the older WinMM names, and use those same names in the older WinRT MIDI API. In addition, we may consider not allowing renaming of MIDI 1.0 device names as they appear in MIDI 1.0 ports (always enumerate with the manufacturer-supplied names in WinMM). Note that this is only for USB MIDI 1.0 byte stream devices.

**Comment by Psychlist1972** [comment]
For MIDI 1.0 devices, this is currently how the naming works.

If there's an iJack value provided, we use that and the filter name <filter> <iJack>

If there's no iJack value provided, we use the filter name <filter> . We may add a differentiator number to that.

The user-provided name always overrides this naming approach, although we do make all names available.

**Comment by tosugawara** [comment]
@Psychlist1972 
I have some questions about this issue:

> For MIDI 1.0 devices, this is currently how the naming works.
> 
> If there's an iJack value provided, we use that and the filter name
> 
> If there's no iJack value provided, we use the filter name . We may add a differentiator number to that.
> 
> The user-provided name always overrides this naming approach, although we do make all names available.

1. On DP7 now,  is the naming for port names for MIDI 1.0 device above working? 
2. When the MIDI 1.0 device having multiple ports is connected, are each iJack values used without vendor specific driver?

**Comment by Psychlist1972** [comment]
@tosugawara In DP7, the naming for the generated Group Terminal Blocks for aggregated MIDI 1.0 ports in the SDK follows the above formula. Currently, the naming of WinMM Ports does not. We're working on it.

One of the challenges with naming is we're limited to a maximum of 32 characters for WinMM MIDI 1.0 Port names and we'll need to include the group number in them.

Did you run into something unexpected?

**Comment by tosugawara** [comment]
> [@tosugawara](https://github.com/tosugawara) In DP7, the naming for the generated Group Terminal Blocks for aggregated MIDI 1.0 ports in the SDK follows the above formula. Currently, the naming of WinMM Ports does not. We're working on it.

Thanks. I thought the GTB names would be also used for WinMM Port names, and I just wanted to confirm this.

**Comment by Psychlist1972** [comment]
> > [@tosugawara](https://github.com/tosugawara) In DP7, the naming for the generated Group Terminal Blocks for aggregated MIDI 1.0 ports in the SDK follows the above formula. Currently, the naming of WinMM Ports does not. We're working on it.
> 
> Thanks. I thought the GTB names would be also used for WinMM Port names, and I just wanted to confirm this.

Yes and no. We have to modify it because the iJack (pin) names are not always helpful. We use the pin names as the port name in WinRT MIDI 1.0. As a result, the names are pretty bad in some cases. For example, the MOTU Express 128 has all pin names as "MIDI" so you end up with a list of ports just named "MIDI". In that case, it would make more sense to also have the filter name as part of the name. In this case, the filter name is 20 characters.

![Image](https://github.com/user-attachments/assets/5890a14f-9014-4450-b1cd-0a9065ca4b46)

Others like the UM-One have no pin names

![Image](https://github.com/user-attachments/assets/33821164-a12d-4c8b-80f6-1698c60708e5)

And then others include the device name in the pin name:

![Image](https://github.com/user-attachments/assets/0f0bc53e-d581-4fc6-abd6-51d592881480)

When you combine all that with the 32 character limitation for WinMM port names, you can see there is no single formula which satisfies all devices and ports. 

So the decision tree for naming is more complex than just using the pin/jack information and includes things like checking to see if the device name is in the pin name, and more.

For native MIDI 2.0 devices the naming is decided in a different way.

**Comment by Psychlist1972** [comment]
The DP8 SDK includes a tool named `midiusbinfo.exe` which displays raw information from our USB stack for kernel streaming devices. From it, you can see the names we see after our stack has processed the headers.

Examples:

![Image](https://github.com/user-attachments/assets/adb344dc-7121-4230-a5bc-8093b25a780f)

On the UM-One, you can see our USB stack auto-generated pin names based on the device name, so we end up having to detect this and ignore these types of names.

![Image](https://github.com/user-attachments/assets/3119e673-eb4a-4a4c-beb5-c25a5221b3fe)

same with the mio

![Image](https://github.com/user-attachments/assets/a28db698-46c1-4a83-b0e8-50b510c44782)

And here's a case where the vendor driver (from MOTU) breaks each in/out pair out into its on filter instead of as separate pins on the same filter, making the physical device show up as 8 logical devices each with a single pair. You can also see that the pin names are not super helpful :)

![Image](https://github.com/user-attachments/assets/5fb37088-f9fe-4a0f-9f0a-18de093ff15b)

That's all to say that we're going to do our best here, but the variation between devices makes this a challenge.

**Comment by tosugawara** [comment]
> That's all to say that we're going to do our best here, but the variation between devices makes this a challenge.

I have understood. Thanks for your detailed explanation.
